README.md
# remotestick-server

## What is it?
RemoteStick server exposes the Tellstick
(see [Telldus Homepage](http://www.telldus.se) ) interface through RESTful
services. RemoteStick server uses the native library libtelldus-core to
communicate with the Tellstick. It supports resource browsing and resource
operations (like turning on/off a device). Responses are given in XML format
(support for JSON format is planned). It aims to fully reflect the capabilities
of the libtelldus-core.

## Prerequisites
telldus-core need be installed. telldus-core is an open-source library developed
by the very same company manufacturing the Tellstick, namely Telldus.
Either:

 * install TelldusCenter from
 [http://download.telldus.se/TellStick/Software/TelldusCenter](http://download.telldus.se/TellStick/Software/TelldusCenter)
 (which will give you the telldus-core library as well) or 
 * compile/install telldus-core manually. telldus-core is found here
 [http://download.telldus.se/TellStick/Software/telldus-core/](http://download.telldus.se/TellStick/Software/telldus-core/)

## Getting it

 * A stable version of RemoteStick Server is found here
 http://github.com/pakerfeldt/remotestick/downloads (recommended).
 * A bleeding edge version of RemoteStick Server can be grabbed from the github
 repository remotestick.

## Using it
The -? flag will give you help about available command line arguments:
    remotestick-server.py -?

Starting RemoteStick server is as simple as (although not recommended, see
below):
    remotestick-server.py

By default, no authentication will be required (making it possible for anyone to
query your tellstick) and the RESTful services will listen for connections on
your-hostname:8422. Only Basic Authentication (HTTP) currently supported.
You should at least set a username and password, requiring client
authentication:
    remotestick-server.py --username MyUsername --password MyPassword

## Troubleshooting

### Linux
Depending on where the telldus-core library is installed on your system you may
have to define the environment variable LD_LIBRARY_PATH to include the directory
where telldus-core is located.
    export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/opt/telldus/bin

### Windows
If TelldusCenter is installed RemoteStick server will have no problem locating
the library. Although you might get an error saying ftdxx library is not found.
In such cases add the Telldus folder to the PATH environment variable.
    set PATH=%PATH%;"C:\Program files\Telldus" 
or try running RemoteStick server using Telldus folder as working directory
(i.e. stand in that directory when starting RemoteStick server).

### Mac OS X
The current version of TelldusCenter for Mac OS X, 2.0.2, does only come with 32
bit support meaning that it will not work (out of the box) if you got a 64 bit
version of Python (which you do if you're running a 64 bit version of Mac OS X,
i.e. Snow Leopard). The error you might get indicating this is:
    OSError: dlopen(/Library/Frameworks/TelldusCore.framework/TelldusCore, 6): no suitable image found.
    Did find:/Library/Frameworks/TelldusCore.framework/TelldusCore: mach-o, but wrong architecture
You have two options. Run python in 32 bit mode (recommended):
    arch -arch i386 python2.6 ./remotestick-server.py ...
OR you could also re-compile a 64 bit version of TelldusCore. Although very much
feasable it is complicated. For instructions, see:
http://www.whyro.org/2010/02/14/java-interface-to-tellstick-in-snow-leopard/

CHANGES.md
# ChangeLog

## 2010-12-31 - pakerfeldt (0.4.1)
 * Fixed bug where -f flag was required for static files

## 2010-12-29 - pakerfeldt (0.4.0)
 * Added support for static files through path /s/

## 2010-08-15 - pakerfeldt (0.3.5)
 * Even better handling of authentication
 * Ignore invalid 'last sent value' result from driver

## 2010-08-02 - pakerfeldt (0.3.4)
 * Correct handling of authentication
 
## 2010-07-25 - pakerfeldt (0.3.3)
 * Fixed issue 11.

## 2010-07-21 - pakerfeldt (0.3.2)
 * Changed default port to 8422.

## 2010-05-18 - pakerfeldt (0.3.1)
 * Added Learn method.

## 2010-05-15 - pakerfeldt (0.3)
 * Fixed bug in XML output

## 2010-05-14 - pakerfeldt (0.2)
 * Fixed library handling to support win32 and linux

## 2010-05-14 - pakerfeldt (0.1)
 * First version released.
 * Supports listing device(s)
 * Supports methods on/off/dim
 * XML format supported
 * Supports deleting/creating devices
 * Rudimentary support for editing devices


remotestick-server/src/RemoteStickServer.py
#!/usr/bin/env python

#
# Copyright (c) 2010, Patrik Akerfeldt. All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without modification, are
# permitted provided that the following conditions are met:
# 
#    1. Redistributions of source code must retain the above copyright notice, this list of
#       conditions and the following disclaimer.
# 
#    2. Redistributions in binary form must reproduce the above copyright notice, this list
#       of conditions and the following disclaimer in the documentation and/or other materials
#       provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY PATRIK AKERFELDT ``AS IS'' AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
# FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL PATRIK AKERFELDT OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
# 
# The views and conclusions contained in the software and documentation are those of the
# authors and should not be interpreted as representing official policies, either expressed
# or implied, of Patrik Akerfeldt.
# 
# 


from bottle import route, run, response, request
from ctypes import cdll, c_char_p, util
from getopt import getopt, GetoptError
from sys import argv, exit, platform
from base64 import b64encode

#Device methods
TELLSTICK_TURNON = 1
TELLSTICK_TURNOFF = 2
TELLSTICK_BELL = 4
TELLSTICK_TOGGLE = 8
TELLSTICK_DIM = 16
TELLSTICK_LEARN = 32
ALL_METHODS = TELLSTICK_TURNON | TELLSTICK_TURNOFF | TELLSTICK_BELL | TELLSTICK_TOGGLE | TELLSTICK_DIM | TELLSTICK_LEARN

#Error codes
TELLSTICK_SUCCESS = 0
TELLSTICK_ERROR_NOT_FOUND = -1
TELLSTICK_ERROR_PERMISSION_DENIED = -2
TELLSTICK_ERROR_DEVICE_NOT_FOUND = -3
TELLSTICK_ERROR_METHOD_NOT_SUPPORTED = -4
TELLSTICK_ERROR_COMMUNICATION = -5
TELLSTICK_ERROR_UNKNOWN = -99

#Device typedef
TELLSTICK_TYPE_DEVICE = 1
TELLSTICK_TYPE_GROUP = 2

#Protocol Nexa
TELLSTICK_DEVICE_YCR3500 = 1
TELLSTICK_DEVICE_YCR300D = 2
TELLSTICK_DEVICE_WSR1000 = 3
TELLSTICK_DEVICE_CMR1000 = 4
TELLSTICK_DEVICE_CMR300 = 5
TELLSTICK_DEVICE_PA33300 = 6
TELLSTICK_DEVICE_EL2000 = 8
TELLSTICK_DEVICE_EL2005 = 9
TELLSTICK_DEVICE_EL2006 = 10
TELLSTICK_DEVICE_SYCR3500 = 12
TELLSTICK_DEVICE_SYCR300 = 13
TELLSTICK_DEVICE_HDR105 = 14
TELLSTICK_DEVICE_ML7100 = 15
TELLSTICK_DEVICE_EL2004 = 16
TELLSTICK_DEVICE_EL2016 = 17
TELLSTICK_DEVICE_EL2010 = 18
TELLSTICK_DEVICE_LYCR1000 = 20
TELLSTICK_DEVICE_LYCR300 = 21
TELLSTICK_DEVICE_LCMR1000 = 22
TELLSTICK_DEVICE_LCMR300 = 23
TELLSTICK_DEVICE_EL2023 = 24
TELLSTICK_DEVICE_EL2024 = 25
TELLSTICK_DEVICE_EL2021 = 26
TELLSTICK_DEVICE_EL2017 = 27
TELLSTICK_DEVICE_EL2019 = 28

#Protocol Ikea
TELLSTICK_DEVICE_KOPPLA = 19

reqauth = True
encodedauth = None
libtelldus = None

def loadlibrary(libraryname=None):
    if libraryname == None and platform == "darwin":
        libraryname = "TelldusCore"
    elif libraryname == None:
        libraryname = "libtelldus-core"
        
    ret = util.find_library(libraryname)
    
    if ret == None:
        return (None, libraryname)

    global libtelldus
    libtelldus = cdll.LoadLibrary(ret)
    libtelldus.tdGetName.restype = c_char_p
    libtelldus.tdLastSentValue.restype = c_char_p
    libtelldus.tdGetProtocol.restype = c_char_p
    libtelldus.tdGetModel.restype = c_char_p
    libtelldus.tdGetErrorString.restype = c_char_p

    return ret, libraryname

def errmsg(x):
    return {
        100: "API key could not be verified",
        101: "Unsupported format",
        201: "Name not supplied",
        202: "Model not supplied",
        203: "Protocol not supplied",
        210: "Malformed parameters",
        211: "No device removed",
        220: "Method not supported",
        300: "Telldus-core error"
    }[x]
   
def err(format, responsecode, request, code, code_msg=None):
    response.status = responsecode
    if code_msg == None:
        code_msg = errmsg(code)
    
    if format == "xml":
        return err_xml(request, code_msg)
    else:
        return err_xml(request, code_msg)

def err_xml(request, msg):
    return "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<hash>\n\t<request>" + request + "</request>\n\t<error>" + msg + "</error>\n</hash>"

def authenticate(encoded_base64):
    global encodedauth
    if reqauth:
        return encodedauth == encoded_base64
    else:
        return True

def read_device(identity):
    name = libtelldus.tdGetName(identity)
    lastcmd = libtelldus.tdLastSentCommand(identity, 1)
    protocol = libtelldus.tdGetProtocol(identity)
    model = libtelldus.tdGetModel(identity)
    methods = libtelldus.tdMethods(identity, ALL_METHODS)
    element = "<device id=\"" + str(identity) + "\">\n\t\t<name>" + name + "</name>\n\t\t<protocol>" + protocol + "</protocol>\n\t\t<model>" + model + "</model>\n"
    element += "\t\t<lastcmd>" + ("ON" if lastcmd == 1 else "OFF") + "</lastcmd>\n"
    if methods & TELLSTICK_BELL:
        element += "\t\t<supportedMethod id=\"" + str(TELLSTICK_BELL) + "\">" + "TELLSTICK_BELL</supportedMethod>\n"
    if methods & TELLSTICK_TOGGLE:
        element += "\t\t<supportedMethod id=\"" + str(TELLSTICK_TOGGLE) + "\">" + "TELLSTICK_TOGGLE</supportedMethod>\n"
    if methods & TELLSTICK_TURNOFF:
        element += "\t\t<supportedMethod id=\"" + str(TELLSTICK_TURNOFF) + "\">" + "TELLSTICK_TURNOFF</supportedMethod>\n"
    if methods & TELLSTICK_TURNON:
        element += "\t\t<supportedMethod id=\"" + str(TELLSTICK_TURNON) + "\">" + "TELLSTICK_TURNON</supportedMethod>\n"
    if methods & TELLSTICK_DIM:
        element += "\t\t<supportedMethod id=\"" + str(TELLSTICK_DIM) + "\">" + "TELLSTICK_DIM</supportedMethod>\n"
    if methods & TELLSTICK_LEARN:
        element += "\t\t<supportedMethod id=\"" + str(TELLSTICK_LEARN) + "\">" + "TELLSTICK_LEARN</supportedMethod>\n"
    element += "</device>\n"
    return element

def pre_check(format, accepted_formats):
    if format not in accepted_formats:
        return False, 400, 101
    
    if not authenticate(request.authorization):
        return False, 401, 100
    
    return True, None, None

def set_content_type(format):
    if format == "xml":
        response.content_type = 'text/xml; charset=UTF-8'

@route('/devices\.:format', method='GET')
def devices(format):
    ok, response_code, error_code = pre_check(format, ["xml"])
    if not ok:
        return err(format, response_code, 'GET /devices.' + format, error_code)
    set_content_type(format)

    result = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<devices>\n"
    numDevices = libtelldus.tdGetNumberOfDevices()
    for i in range(numDevices):
        result += read_device(libtelldus.tdGetDeviceId(i))
    result += "</devices>"
    return result

@route('/devices\.:format', method='POST')
def new_device(format):
    request_str = 'POST /devices.' + format
    ok, response_code, error_code = pre_check(format, ["xml"])
    if not ok:
        return err(format, response_code, request_str, error_code)
    set_content_type(format)
    
    name = request.POST.get('name', '').strip()
    if not name:
        return err(format, 400, request_str, 201)

    model = request.POST.get('model', '').strip()
    if not model:
        return err(format, 400, request_str, 202)

    protocol = request.POST.get('protocol', '').strip()
    if not protocol:
            return err(format, 400, request_str, 203)
        
    rawParams = request.POST.get('parameters', '').strip()
    print rawParams
    parameters = []
    if rawParams != None:
        for param in rawParams.split():
            keyval = param.split('=')
            if len(keyval) != 2:
                return err(format, 400, request_str, 210)
            else:
                parameters.append(keyval)

    identity = libtelldus.tdAddDevice()
    libtelldus.tdSetName(identity, name)
    libtelldus.tdSetProtocol(identity, protocol)
    libtelldus.tdSetModel(identity, model)
    print parameters
    for param in parameters:
        libtelldus.tdSetDeviceParameter(identity, param[0], param[1])

    retval = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
    retval += read_device(identity)
    return retval

@route('/devices/:id\.:format', method='GET')
def get_device(id, format):
    request_str = 'GET /devices/' + id + "." + format
    ok, response_code, error_code = pre_check(format, ["xml"])
    if not ok:
        return err(format, response_code, request_str, error_code)
    set_content_type(format)

    retval = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
    try:
        retval += read_device(int(id))
        return retval
    except ValueError:
        return err(format, 400, request_str, 210)


@route('/devices/:id\.:format', method='DELETE')
def delete_device(id, format):
    request_str = 'DELETE /devices/' + id + "." + format
    ok, response_code, error_code = pre_check(format, ["xml"])
    if not ok:
        return err(format, response_code, request_str, error_code)
    set_content_type(format)

    try:
        retval = libtelldus.tdRemoveDevice(int(id))
    except ValueError:
        return err(format, 400, request_str, 210)

    if retval == 1:
        return ""
    else:
        return err(format, 400, request_str, 211)

@route('/devices/:id\.:format', method='PUT')
def change_device(id, format):
    request_str = 'PUT /devices/' + id + "." + format
    ok, response_code, error_code = pre_check(format, ["xml"])
    if not ok:
        return err(format, response_code, request_str, error_code)
    set_content_type(format)

    name = request.POST.get('name', '').strip()
    protocol = request.POST.get('protocol', '').strip()
    model = request.POST.get('model', '').strip()
    if not name:
        return err(format, 400, request_str, 101)
        
    libtelldus.tdSetName(int(id), name)
    libtelldus.tdSetProtocol(int(id), protocol)
    libtelldus.tdSetModel(int(id), model)
    
    retval = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
    try:
        retval += read_device(int(id))
        return retval
    except ValueError:
        return err(format, 400, request_str, 210)
    return ""

@route('/devices/:id/on\.:format', method='GET')
def turnon_device(id, format):
    request_str = 'GET /devices/' + id + "/on." + format
    ok, response_code, error_code = pre_check(format, ["xml"])
    if not ok:
        return err(format, response_code, request_str, error_code)
    set_content_type(format)
    
    try:
        identity = int(id)
    except ValueError:
        return err(format, 400, request_str, 210)
    
    if libtelldus.tdMethods(identity, TELLSTICK_TURNON) & TELLSTICK_TURNON:    
        retval = libtelldus.tdTurnOn(identity)
        if retval == 0:
            return ""
        else:
            return err(format, 502, request_str, 300, libtelldus.tdGetErrorString(retval))
    else:
        return err(format, 400, request_str, 220)

@route('/devices/:id/off\.:format', method='GET')
def turnoff_device(id, format):
    request_str = 'GET /devices/' + id + "/off." + format
    ok, response_code, error_code = pre_check(format, ["xml"])
    if not ok:
        return err(format, response_code, request_str, error_code)
    set_content_type(format)

    try:
        identity = int(id)
    except ValueError:
        return err(format, 400, request_str, 210)
    
    if libtelldus.tdMethods(identity, TELLSTICK_TURNOFF) & TELLSTICK_TURNOFF:    
        retval = libtelldus.tdTurnOff(identity)
        if retval == 0:
            return ""
        else:
            return err(format, 502, request_str, 300, libtelldus.tdGetErrorString(retval))
    else:
        return err(format, 400, request_str, 220)

@route('/devices/:id/dim/:level\.:format', method='GET')
def dim_device(id, level, format):
    request_str = 'GET /devices/' + id + "/dim/" + level + "." + format
    ok, response_code, error_code = pre_check(format, ["xml"])
    if not ok:
        return err(format, response_code, request_str, error_code)
    set_content_type(format)

    try:
        identity = int(id)
        dimlevel = chr(int(level))
    except ValueError:
        return err(format, 400, request_str, 210)

    if libtelldus.tdMethods(identity, TELLSTICK_DIM) & TELLSTICK_DIM:    
        retval = libtelldus.tdDim(identity, dimlevel)
        if retval == 0:
            return ""
        else:
            return err(format, 502, request_str, 300, libtelldus.tdGetErrorString(retval))
    else:
        return err(format, 400, request_str, 220)

def usage():
    print "Usage: RemoteStickServer [OPTION]..."
    print "Expose tellstick services through RESTful services."
    print ""
    print "Without any arguments RemoteStickServer will start a http server on 127.0.0.1:8080 where no authentication is required."
    print "Setting the name of the telldus-core library should not be needed. RemoteStickServer is able to figure out the correct library name automatically. If, for some reason, this is unsuccessful, use --library."
    print ""
    print "-h, --host\t\thost/IP which the server will bind to, default to loopback"
    print "-p, --post\t\tport which the server will listen on, default to 8080"
    print "-u, --username\t\tusername used for client authentication"
    print "-p, --password\t\tpassword used for client authentication"
    print "-l, --library\t\tname of telldus-core library"

def main():
    try:
        opts, args = getopt(argv[1:], "?h:p:u:s:l:", ["?", "host=", "port=", "username=", "password=", "library="])
    except GetoptError, err:
        print str(err)
        usage()
        exit(2)
    host = None
    port = None
    library = None
    username = None
    password = None
    global reqauth
    global encodedauth
    
    for o, a in opts:
        if o in ("-h", "--host"):
            host = a
        elif o in ("-p", "--port"):
            port = a
        elif o in ("-u", "--username"):
            username = a
        elif o in ("-s", "--password"):
            password = a
        elif o in ("-l", "--library"):
            library = a
        elif o == '-?':
            usage()
            exit()
        else:
            assert False, "unhandled option " + o
    
    lib, libname = loadlibrary(library)
    if lib == None:
        print "Error: Cannot find library " + libname
        
    if username == None or password == None:
        print "Warning: No authentication required. Please consider setting --username and --password."
        reqauth = False
    else:
        encodedauth = "Basic " + b64encode(username + ":" + password)
        
    if (host == None and port == None):
        run()
    elif host != None and port == None:
        run(host=host)
    elif host == None and port != None:
        run(port=port)
    else:
        run(host=host, port=port)
             
if __name__ == "__main__":
    main()
